### PR TITLE
Add Drop out

### LINKED ISSUE #36: Implement Drop Out

### PR DESCRIPTION
**Pull request by Hattomo**
## Description

Add drop out
## Related Issues and Reference

Close #36

### CODE REVIEW

**Review by kaikai529 — APPROVED** [review]

**Comment by Hattomo** [comment]
@kaikai529 I change a lot of code since last you visit, so can you review again?

File: dataset/logic_circuit.py [hunk 1]
```diff
@@ -86,6 +86,20 @@ def dset(d_name, num):
...
+            data[4*i + 1] = [255, 255, 255, 255, 255, 0, 0, 255, 255, 0, 0, 255, 255, 255, 255, 255]
+            data[4*i + 2] = [0, 0, 255, 0, 255, 255, 255, 255, 0, 0, 255, 0, 0, 0, 255, 0]
+            data[4*i + 3] = [0, 0, 0, 0, 0, 255, 255, 0, 0, 255, 255, 0, 0, 0, 0, 255]
+        label = np.zeros((4 * num, 4))
+        for i in range(num):
+            label[4 * i] = [1, 0, 0, 0]
+            label[4*i + 1] = [0, 1, 0, 0]
+            label[4*i + 2] = [0, 0, 1, 0]
+            label[4*i + 3] = [0, 0, 0, 1]
+        return data, label
```
Comment (kaikai529) [inline dataset/logic_circuit.py:102]: Image data is (channel, height , width).
If Image data is grayscale , channel is 1.

"""
data = np.zeros([4*num (number of data),1(channel), 4(data height), 4(data width)]) 
"""

File: dnn/neural_network.py [hunk 2]
```diff
@@ -65,14 +67,20 @@ def __set_costfunc(self, costfunc):
...
+            self.y[i + 1][1:] = self.weight[i] @ self.do[i] @ self.z[i]
+            if istrain:
+                self.z[i + 1] = self.actfunc(self.y[i + 1])
+            else:
+                self.z[i + 1] = self.actfunc(self.y[i + 1]) * (1 - self.dropout[i])
+        self.y[-1] = self.weight[-1] @ self.do[-1] @ self.z[-2]
+        if istrain:
+            self.z[-1] = self.actfunc(self.y[-1])
+        else:
+            self.z[-1] = self.actfunc(self.y[-1]) * (1 - self.dropout[i])
```
Comment (kaikai529) [inline dnn/neural_network.py]: 83  : self.z[-1] = self.actfunc(self.y[-1]) * (1 - self.dropout[i])

self.dropout[i] ?
here is out of for.

File: dnn/main.py [hunk 3]
```diff
@@ -8,16 +8,17 @@
 import numpy_files as npfiles
 import logic_circuit as lc
 
-structure = [3, 3, 1]
-epoch = 30
+structure = [3, 3, 2]
+dropout = [0, 0, 0]
```
Comment (kaikai529) [inline dnn/main.py:12]: dropout is [0, 0, 0] but what does Last element do ?

File: dnn/neural_network.py [hunk 4]
```diff
@@ -65,14 +67,20 @@ def __set_costfunc(self, costfunc):
...
         for i in range(len(self.structure) - 2):
-            self.y[i + 1][1:] = self.weight[i] @ self.z[i]
-            self.z[i + 1] = self.actfunc(self.y[i + 1])
-        self.y[-1] = self.weight[-1] @ self.z[-2]
-        self.z[-1] = self.actfunc(self.y[-1])
+            self.y[i + 1][1:] = self.weight[i] @ self.do[i] @ self.z[i]
+            if istrain:
+                self.z[i + 1] = self.actfunc(self.y[i + 1])
+            else:
+                self.z[i + 1] = self.actfunc(self.y[i + 1]) * (1 - self.dropout[i])
```
Comment (kaikai529) [inline dnn/neural_network.py:78]: what does this codes do ?

**Review by kaikai529 — CHANGES_REQUESTED** [review]

File: dataset/logic_circuit.py (hunk 1 above)
Comment (Hattomo) [reply dataset/logic_circuit.py:102]: this is not image data.
this is 1*16 numpy for input layer

File: dnn/neural_network.py (hunk 2 above)
Comment (Hattomo) [reply dnn/neural_network.py]: Thanks!

File: dataset/logic_circuit.py (hunk 1 above)
Comment (kaikai529) [reply dataset/logic_circuit.py:102]: Ok

**Review by kaikai529 — APPROVED** [review]

**Pull request merged**